plugins/kubernetes/app/models/kubernetes/build_job_executor.rb

```diff
@@ -0,0 +1,140 @@
+      end
+      @registry = registry
+
+      # Read the external config path and create a new job config instance
+      contents = File.read(build_job_config_path)
+      k8s_job = Kubernetes::RoleConfigFile.new(contents, build_job_config_path).job
+      fill_job_details(k8s_job, build, project, tag: tag, push: push, tag_as_latest: tag_as_latest)
+
+      if k8s_job
+        job_name = k8s_job[:metadata][:name]
+        job_namespace = k8s_job[:metadata][:namespace]
+      end
+
+      @output.puts "### Running a remote Docker build job: #{job_name}"
+      success, job_log = create_and_wait_for_job(k8s_job, @output)
+
+      if success
+        @output.puts "### Remote build job #{job_name} completed sucessfully"
+      else
+        @output.puts "### Remote build job #{job_name} failed to run or timed out"
+      end
+
+      return success, job_log
+    ensure
+      if k8s_job
+        # Jobs will still be there regardless of their statuses
+        @output.puts "### Cleaning up the remote build job: #{job_name}"
+        extension_client.delete_job(job_name, job_namespace)
+      end
+    end
+
+    private
+
+    def build_cluster
+      Kubernetes::Cluster.first!
+    end
+
+    def client
+      build_cluster.client
+    end
+
+    def extension_client
+      build_cluster.extension_client
+    end
+
+    def valid_registry_info?(registry)
+      registry[:serveraddress].present?
+    end
+
+    def build_job_config_path
+      ENV['KUBE_BUILD_JOB_FILE'] || File.join(Rails.root, 'plugins', 'kubernetes', 'config', 'build_job.yml')
+    end
+
+    def fill_job_details(k8s_job, build, project, tag:, push: false, tag_as_latest: false)
+      # Fill in some information to easily query the job resource
+      project_name = project.permalink.tr('_', '-')
+      labels = { project: project_name, role: "docker-build-job" }
+      k8s_job[:metadata][:name] = "#{project_name}-docker-build-#{build.id}-#{SecureRandom.hex(7)}"
+      k8s_job[:metadata][:labels].update(labels)
+      k8s_job[:spec][:template][:metadata][:labels].update(labels)
+
+      # Necessary to query the pod running the job later
+      k8s_job[:spec][:template][:metadata][:labels][:job_name] = k8s_job[:metadata][:name]
+
+      # Pass all necessary information so that remote container can build the image
+      container_params = {
+        env: [{name: 'DOCKER_REGISTRY', value: @registry[:serveraddress] }],
+        args: [
+          project.repository_url, build.git_sha, project.docker_repo, tag,
+          push ? "yes" : "no",
+          tag_as_latest ? "yes" : "no"
+        ]
+      }
+      k8s_job[:spec][:template][:spec][:containers][0].update(container_params)
+    end
+
+    def create_and_wait_for_job(k8s_job, output)
+      extension_client.create_job(k8s_job.deep_symbolize_keys)
+      start = Time.now
+      logs = pod_name = ""
+      job_name = k8s_job[:metadata][:name]
+      job_namespace = k8s_job[:metadata][:namespace]
+
+      loop do
```

Comment:
not asking about containers, but why we need to loop here ... I was thinking we'd only have 1 pod created / don't need any looping logic